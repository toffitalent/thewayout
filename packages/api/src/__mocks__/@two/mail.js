const send = jest.fn().mockResolvedValue(undefined);

module.exports = {
  send,
};
