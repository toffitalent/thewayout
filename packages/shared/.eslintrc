{
  "settings": {
    "import/resolver": {
      "typescript": {
        "project": ["tsconfig.json", "packages/shared/tsconfig.json"]
      }
    }
  },
  "rules": {
    "no-param-reassign": "off",
    "react/react-in-jsx-scope": "off"
  }
}
