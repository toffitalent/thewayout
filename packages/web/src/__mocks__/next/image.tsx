import * as React from 'react';

// eslint-disable-next-line react/jsx-no-useless-fragment
const mock = (props: { children: React.ReactElement }): React.ReactElement => <>{props.children}</>;

export default mock;
