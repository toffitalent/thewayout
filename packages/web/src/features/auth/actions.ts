import { CreateUserRequest } from '@two/shared';
import { API } from '@app/api';
import { createApiThunk } from '@app/utils';

export const fetchUser = createApiThunk('auth/fetchUser', API.user.retrieve);
export const forgotPassword = createApiThunk('auth/forgotPassword', API.user.forgotPassword);

export const login = createApiThunk(
  'auth/login',
  async (credentials: Parameters<typeof API.login>[0], { dispatch }) => {
    await API.login(credentials);
    await dispatch(fetchUser());
  },
);

export const logout = createApiThunk('auth/logout', API.logout);

export const resetPassword = createApiThunk(
  'auth/resetPassword',
  async (
    { username, password, token }: { username: string; password: string; token: string },
    { dispatch },
  ) => {
    await API.user.resetPassword(password, token);
    return dispatch(login({ username, password }));
  },
);

export const signUp = createApiThunk(
  'auth/register',
  async (payload: CreateUserRequest, { dispatch }) => {
    await API.user.create(payload);
    return dispatch(login({ username: payload.email, password: payload.password }));
  },
);

export const updateUser = createApiThunk('auth/updateUser', API.user.update);
