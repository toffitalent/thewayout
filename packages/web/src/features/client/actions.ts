import { ListRspClientRequest } from '@two/shared';
import { API } from '@app/api';
import type { State } from '@app/store';
import { createApiThunk } from '@app/utils';

export const createClient = createApiThunk('client/create', API.clientProfile.create);

export const updateClient = createApiThunk('client/update', API.clientProfile.update);

export const listAllJobs = createApiThunk(
  'client/jobs',
  ({ forceReload }: { forceReload?: boolean }, { getState }) => {
    const { ids } = (getState() as State).clientJobs.jobsList;
    const lastId = [...ids].pop() as string;
    return API.job.list({
      ...(!forceReload && { before: lastId }),
    });
  },
);

export const retrieveJob = createApiThunk(
  'client/job',
  ({ jobId, since }: { jobId: string; since?: Date }) =>
    API.job.retrieve(jobId, {
      ...(since ? { since } : {}),
    }),
);

export const applyJob = createApiThunk(
  'client/jobApply',
  ({ jobId, questions }: { jobId: string; questions?: { [key: string]: string }[] }) =>
    API.job.apply(jobId, {
      ...(questions ? { questions } : {}),
    }),
);

export const listAppliedJobs = createApiThunk(
  'client/applications',
  ({ forceReload, userId }: { forceReload?: boolean; userId: string }, { getState }) => {
    const { ids } = (getState() as State).clientJobs.appliedJobs;
    const lastId = [...ids].pop() as string;
    return API.clientProfile.listJobs(userId, {
      ...(!forceReload && { before: lastId }),
    });
  },
);

export const listSuggestedJobs = createApiThunk('client/suggestedJobs', (clientId: string) =>
  API.clientProfile.listSuggestedJobs(clientId),
);

export const listRsp = createApiThunk(
  'rsp/list',
  ({ userId, ...rest }: ListRspClientRequest & { userId: string }) =>
    API.clientProfile.listRsp({ userId, ...rest }),
);

export const retrieveCaseManager = createApiThunk(
  'client/caseManager',
  API.clientProfile.getCaseManager,
);
