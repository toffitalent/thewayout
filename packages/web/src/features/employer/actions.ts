import { Job, JobApplicationStatus } from '@two/shared';
import { API } from '@app/api';
import type { State } from '@app/store';
import { createApiThunk } from '@app/utils';

export const createEmployer = createApiThunk('employer/create', API.employer.create);

export const updateEmployer = createApiThunk('employer/update', API.employer.update);

export const createJob = createApiThunk('employer/job/create', API.job.create);

export const listJobs = createApiThunk(
  'employer/jobs',
  ({ forceReload, userId }: { forceReload?: boolean; userId: string }, { getState }) => {
    const { ids } = (getState() as State).employerJobs.jobs;
    const lastId = [...ids].pop() as string;
    return API.employer.listJobs(userId, {
      ...(!forceReload && { before: lastId }),
    });
  },
);

export const updateListJobs = createApiThunk(
  'employer/jobs/listUpdated',
  ({ userId }: { userId: string }, { getState }) => {
    const { listUpdatedAt, ids } = (getState() as State).employerJobs.jobs;
    const firstId = ids[0] as string;
    const lastId = [...ids].pop() as string;
    return API.employer.listJobs(userId, {
      since: listUpdatedAt,
      after: firstId,
      before: lastId,
    });
  },
);

export const retrieveJob = createApiThunk(
  'employer/job',
  ({ jobId, since }: { jobId: string; since?: Date }) =>
    API.job.retrieve(jobId, {
      ...(since ? { since } : {}),
    }),
);

export const listApplications = createApiThunk(
  'employer/listApplications',
  ({ employerId, jobId }: { employerId: string; jobId: string }, { getState }) => {
    const { entities } = (getState() as State).employerJobs.jobsApplications;
    const page = entities[jobId]?.page;
    const isListEmpty = !entities[jobId]?.applications.length;

    return API.employer.listApplications(employerId, jobId, {
      page: isListEmpty ? 0 : page! + 1,
    });
  },
);

export const getClient = createApiThunk(
  'employer/getClient',
  ({ employerId, jobId, clientId }: { employerId: string; jobId: string; clientId: string }) =>
    API.employer.getClient(employerId, jobId, clientId, {}),
);

export const updateJob = createApiThunk(
  'employer/updateJob',
  ({ jobId, patch }: { jobId: string; patch: Partial<Job> }) => API.job.update(jobId, patch),
);

export const interviewClient = createApiThunk(
  'employer/interviewClient',
  ({
    employerId,
    jobId,
    applicationId,
  }: {
    employerId: string;
    jobId: string;
    applicationId: string;
  }) =>
    API.employer.changeApplicationStatus(
      employerId,
      jobId,
      applicationId,
      JobApplicationStatus.interview,
    ),
);

export const hireClient = createApiThunk(
  'employer/hireClient',
  ({
    employerId,
    jobId,
    applicationId,
  }: {
    employerId: string;
    jobId: string;
    applicationId: string;
  }) =>
    API.employer.changeApplicationStatus(
      employerId,
      jobId,
      applicationId,
      JobApplicationStatus.hired,
    ),
);

export const rejectClient = createApiThunk(
  'employer/rejectClient',
  ({
    employerId,
    jobId,
    applicationId,
  }: {
    employerId: string;
    jobId: string;
    applicationId: string;
  }) =>
    API.employer.changeApplicationStatus(
      employerId,
      jobId,
      applicationId,
      JobApplicationStatus.rejected,
    ),
);

export const notAFitClient = createApiThunk(
  'employer/notAFitClient',
  ({
    employerId,
    jobId,
    applicationId,
  }: {
    employerId: string;
    jobId: string;
    applicationId: string;
  }) =>
    API.employer.changeApplicationStatus(
      employerId,
      jobId,
      applicationId,
      JobApplicationStatus.notAFit,
    ),
);
