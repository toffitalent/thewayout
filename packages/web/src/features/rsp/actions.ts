import { ClientType, RspClientStatus } from '@two/shared';
import { API } from '@app/api';
import type { State } from '@app/store';
import { createApiThunk } from '@app/utils';

export const createRsp = createApiThunk('rsp/create', API.rsp.create);

export const updateRsp = createApiThunk('rsp/update', API.rsp.update);

export const createMember = createApiThunk('rsp/createMember', API.rsp.createMember);

export const inviteCaseManager = createApiThunk('rsp/inviteCaseManager', API.rsp.inviteCaseManager);

export const removeInvitation = createApiThunk('rsp/removeInvitation', API.rsp.removeInvitation);

export const removeCaseManager = createApiThunk('rsp/removeCaseManager', API.rsp.removeCaseManager);

export const listCaseManagers = createApiThunk(
  'rsp/caseManagers',
  ({ rspId }: { rspId: string }, { getState }) => {
    const { ids } = (getState() as State).rsp.caseManagers;
    const lastId = [...ids].pop() as string;
    return API.rsp.listCaseManagers(rspId, {
      before: lastId,
    });
  },
);

export const listInvitations = createApiThunk(
  'rsp/invitations',
  ({ rspId }: { rspId: string }, { getState }) => {
    const { ids } = (getState() as State).rsp.invitations;
    const lastId = [...ids].pop() as string;
    return API.rsp.listInvitations(rspId, {
      before: lastId,
    });
  },
);

export const listCaseManagersNames = createApiThunk(
  'rsp/listCaseManagersNames',
  API.rsp.listCaseManagersNames,
);

export const retrieveRspAccount = createApiThunk('rsp/retrieveRspAccount', API.rsp.retrieveAccount);

export const updateCaseManager = createApiThunk('rsp/updateCaseManager', API.rsp.updateCaseManager);

export const listClients = createApiThunk(
  'rsp/clients',
  ({ rspId }: { rspId: string }, { getState }) => {
    const { ids } = (getState() as State).rsp.clients;
    const lastId = [...ids].pop() as string;
    return API.rsp.listClients(rspId, {
      before: lastId,
    });
  },
);

export const listStatisticsClients = createApiThunk(
  'rsp/statisticsClients',
  (
    {
      rspId,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      forceReload,
      ...clientFilters
    }: Partial<ClientType> & { rspId: string; forceReload?: boolean },
    { getState },
  ) => {
    const { ids } = (getState() as State).rsp.statisticsClients;
    const lastId = [...ids].pop() as string;
    return API.rsp.listClients(rspId, {
      ...clientFilters,
      before: lastId,
    });
  },
);

export const retrieveClient = createApiThunk('rsp/retrieveClient', API.rsp.retrieveClient);

export const declineClient = createApiThunk(
  'rsp/declineClient',
  ({ rspId, clientId }: { rspId: string; clientId: string }) =>
    API.rsp.changeClientStatus(rspId, clientId, RspClientStatus.closed),
);

export const acceptClient = createApiThunk(
  'rsp/acceptClient',
  ({ rspId, clientId }: { rspId: string; clientId: string }) =>
    API.rsp.changeClientStatus(rspId, clientId, RspClientStatus.active),
);

export const addClientNotes = createApiThunk('rsp/addClientNotes', API.rsp.addClientNotes);

export const assignCaseManager = createApiThunk(
  'rsp/assignCaseManager',
  ({
    rspId,
    clientId,
    caseManagerId,
  }: {
    rspId: string;
    clientId: string;
    caseManagerId: string;
  }) => API.rsp.assignCaseManager(rspId, clientId, caseManagerId),
);

export const getStatistics = createApiThunk('rsp/statistics', API.rsp.getStatistics);
