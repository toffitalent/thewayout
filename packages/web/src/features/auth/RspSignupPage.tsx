import { Box, Button, Field, Flex, Heading, Text, useForm } from '@disruptive-labs/ui';
import Link from 'next/link';
import { useRouter } from 'next/router';
import { useEffect } from 'react';
import { CreateUserRequest } from '@two/shared';
import { UserType } from '@app/api';
import { SEO } from '@app/components/SEO';
import { useAppDispatch, useAuthRedirect, useErrorIndicator } from '@app/hooks';
import { NextPageWithLayout } from '@app/types';
import { signUp } from './actions';
import { AuthLayout } from './AuthLayout';

export const RspSignupPage: NextPageWithLayout = function RspSignupPage() {
  const dispatch = useAppDispatch();
  const showError = useErrorIndicator({
    ResourceConflict: 'Email already in use',
    ResourceNotFound: 'Invitation not found',
  });
  const redirect = useAuthRedirect();
  const router = useRouter();
  const { firstName, lastName, email, phone, invitationId } = router.query as {
    firstName: string;
    lastName: string;
    email: string;
    phone: string;
    invitationId: string;
  };

  const {
    control,
    submitForm,
    setValue,
    formState: { isSubmitting, isDirty },
  } = useForm({
    defaultValues: { firstName: '', lastName: '', email: '', password: '' },
    onSubmit: (values: Omit<CreateUserRequest, 'type'>) =>
      dispatch(
        signUp({
          ...values,
          type: UserType.Rsp,
          phone,
          ...(invitationId && { invitationId }),
        }),
      )
        .unwrap()
        .then(() => redirect())
        .catch((e) => {
          showError(e);
        }),
  });

  const navigateToLogin = () =>
    router.replace({
      pathname: '/login/',
      query: router.query.return_to
        ? {
            return_to: router.query.return_to,
          }
        : undefined,
    });

  useEffect(() => {
    if (firstName) setValue('firstName', firstName);
    if (lastName) setValue('lastName', lastName);
    if (email) setValue('email', email);
  }, [email, setValue, firstName, lastName]);

  return (
    <>
      <SEO title="Sign Up" description="" noIndex />
      <Heading as="h1" f="h2" mb={2.5}>
        Sign Up
      </Heading>
      <Text mb={10}>
        Welcome to The Way Out. Sign up with an email and password to get started on your journey
        today.
      </Text>
      <form onSubmit={submitForm}>
        <Field
          type="email"
          name="email"
          label="Email"
          placeholder="Your email address"
          validate={{ email: true }}
          fluid
          control={control}
          disabled={!!email}
        />
        <Field
          type="password"
          name="password"
          label="Password"
          placeholder="Enter your password"
          validate={{ minLength: 8 }}
          fluid
          control={control}
        />
        <Flex direction="row" mv={4}>
          <Field
            type="text"
            name="firstName"
            label="First Name"
            placeholder="Your first name"
            validate={{ required: true }}
            autoCapitalize="words"
            width="1/2"
            mr={4}
            control={control}
            disabled={!!firstName}
          />
          <Field
            type="text"
            name="lastName"
            label="Last Name"
            placeholder="Your last name"
            validate={{ required: true }}
            autoCapitalize="words"
            width="1/2"
            mt={0}
            control={control}
            disabled={!!lastName}
          />
        </Flex>
        <Field
          type="checkbox"
          name="terms"
          label={
            <>
              {'By signing up, I agree to the '}
              <Link href="/terms/" target="_blank" rel="noopener noreferrer">
                <Text display="inline" fontWeight="700">
                  Terms of Service
                </Text>
              </Link>
              {' and '}
              <Link href="/privacy/" target="_blank" rel="noopener noreferrer">
                <Text display="inline" fontWeight="700">
                  Privacy Policy
                </Text>
              </Link>
              .
            </>
          }
          validate={{ required: true }}
          control={control}
        />
        <Button
          mt={10}
          type="submit"
          colorScheme="primary"
          disabled={isSubmitting || !isDirty}
          fluid
        >
          Sign Up
        </Button>
      </form>
      <Box mt={5} onClick={navigateToLogin}>
        {'Already using The Way Out? '}
        <Text as="a" fontWeight="700">
          Log In
        </Text>
      </Box>
    </>
  );
};

RspSignupPage.getLayout = (page) => <AuthLayout>{page}</AuthLayout>;
