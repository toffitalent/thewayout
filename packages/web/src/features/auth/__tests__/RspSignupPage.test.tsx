import userEvent from '@testing-library/user-event';
import { useRouter } from 'next/router';
import { UserType } from '@two/shared';
import { mockThunks, render } from '@test';
import * as actions from '../actions';
import { RspSignupPage } from '../RspSignupPage';

describe('RspSignupPage', () => {
  beforeEach(() => {
    (useRouter as jest.Mock).mockImplementation(() => ({
      isReady: true,
      query: {},
      replace: jest.fn(),
    }));
  });

  test('renders correctly', async () => {
    const { container } = render(<RspSignupPage />);
    expect(container).toMatchSnapshot();
  });

  test('handles submission', async () => {
    const [signUp] = mockThunks(actions, ['signUp']);
    const { dispatch, getByRole, getByLabelText } = render(<RspSignupPage />);

    await userEvent.type(getByLabelText('First Name'), 'firstname', { delay: 1 });
    await userEvent.type(getByLabelText('Last Name'), 'lastname', { delay: 1 });
    await userEvent.type(getByLabelText('Email'), 'email@email.com', { delay: 1 });
    await userEvent.type(getByLabelText('Password'), 'password', { delay: 1 });
    await userEvent.click(getByLabelText(/^By signing up, I agree to the /));
    await userEvent.click(getByRole('button'));

    expect(signUp).toBeCalledWith({
      firstName: 'firstname',
      lastName: 'lastname',
      email: 'email@email.com',
      password: 'password',
      terms: true,
      type: UserType.Rsp,
    });
    expect(dispatch).toBeCalled();
  });

  test('handles submission by invited case manager', async () => {
    (useRouter as jest.Mock).mockImplementation(() => ({
      isReady: true,
      query: {
        firstName: 'firstname',
        lastName: 'lastname',
        email: 'email@email.com',
        invitationId: 'TEST_ID',
      },
      replace: jest.fn(),
    }));
    const [signUp] = mockThunks(actions, ['signUp']);
    const { dispatch, getByRole, getByLabelText } = render(<RspSignupPage />);

    expect(getByLabelText('First Name')).toBeDisabled();
    expect(getByLabelText('First Name')).toHaveValue('firstname');
    expect(getByLabelText('Last Name')).toBeDisabled();
    expect(getByLabelText('Last Name')).toHaveValue('lastname');
    expect(getByLabelText('Email')).toBeDisabled();
    expect(getByLabelText('Email')).toHaveValue('email@email.com');

    await userEvent.type(getByLabelText('Password'), 'password', { delay: 1 });
    await userEvent.click(getByLabelText(/^By signing up, I agree to the /));
    await userEvent.click(getByRole('button'));

    expect(signUp).toBeCalledWith({
      firstName: 'firstname',
      lastName: 'lastname',
      email: 'email@email.com',
      password: 'password',
      terms: true,
      type: UserType.Rsp,
      invitationId: 'TEST_ID',
    });
    expect(dispatch).toBeCalled();
  });
});
