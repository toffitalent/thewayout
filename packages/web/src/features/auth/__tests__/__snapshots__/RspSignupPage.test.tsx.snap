// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RspSignupPage renders correctly 1`] = `
<div>
  <h1
    class="text marginBottom-2-5 font-h2"
  >
    Sign Up
  </h1>
  <p
    class="text marginBottom-10"
  >
    Welcome to The Way Out. Sign up with an email and password to get started on your journey today.
  </p>
  <form>
    <div
      class="control"
      role="group"
    >
      <label
        class="label"
        for=":r0:"
      >
        Email
      </label>
      <div
        class="root primary md fluid md"
      >
        <input
          class="input"
          id=":r0:"
          name="email"
          placeholder="Your email address"
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      class="control"
      role="group"
    >
      <label
        class="label"
        for=":r1:"
      >
        Password
      </label>
      <div
        class="root primary md fluid md"
      >
        <input
          class="input"
          id=":r1:"
          name="password"
          placeholder="Enter your password"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="flexDirection-row marginTop-4 marginBottom-4"
    >
      <div
        class="control width-1/2 marginRight-4"
        role="group"
      >
        <label
          class="label"
          for=":r2:"
        >
          First Name
        </label>
        <div
          class="root primary md md"
        >
          <input
            autocapitalize="words"
            class="input"
            id=":r2:"
            name="firstName"
            placeholder="Your first name"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="control width-1/2 marginTop-0"
        role="group"
      >
        <label
          class="label"
          for=":r3:"
        >
          Last Name
        </label>
        <div
          class="root primary md md"
        >
          <input
            autocapitalize="words"
            class="input"
            id=":r3:"
            name="lastName"
            placeholder="Your last name"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="control"
      role="group"
    >
      <label
        class="checkbox primary md"
      >
        <input
          class="input"
          id=":r4:"
          name="terms"
          type="checkbox"
          value=""
        />
        <div
          aria-hidden="true"
          class="control borderColor-primary bgcolor-primary"
        >
          <svg
            class="icon"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <polygon
              fill="currentColor"
              points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
            />
          </svg>
        </div>
        <div
          class="label"
        >
          By signing up, I agree to the 
          <a
            href="/terms"
            rel="noopener noreferrer"
            target="_blank"
          >
            <p
              class="text display-inline fontWeight-700"
            >
              Terms of Service
            </p>
          </a>
           and 
          <a
            href="/privacy"
            rel="noopener noreferrer"
            target="_blank"
          >
            <p
              class="text display-inline fontWeight-700"
            >
              Privacy Policy
            </p>
          </a>
          .
        </div>
      </label>
    </div>
    <button
      class="button md solid c-primary disabled fluid marginTop-10"
      disabled=""
      type="submit"
    >
      Sign Up
    </button>
  </form>
  <div
    class="marginTop-5"
  >
    Already using The Way Out? 
    <a
      class="text fontWeight-700"
    >
      Log In
    </a>
  </div>
</div>
`;
